Content.Server/UserInterface/ActivatableUIComponent.cs

```diff
@@ -0,0 +1,95 @@
+{
+    [RegisterComponent]
+    public class ActivatableUIComponent : Component,
+            ISerializationHooks
+    {
+        public override string Name => "ActivatableUI";
+
+        private ActivatableUISystem _activatableUISystem = default!;
+
+        [ViewVariables]
+        public Enum? Key { get; set; }
+
+        [ViewVariables] public BoundUserInterface? UserInterface => (Key != null) ? Owner.GetUIOrNull(Key) : null;
+
+        [ViewVariables(VVAccess.ReadWrite)]
+        [DataField("inHandsOnly")]
+        public bool InHandsOnly { get; set; } = false;
+
+        [ViewVariables]
+        [DataField("singleUser")]
+        public bool SingleUser { get; set; } = false;
+
+        [ViewVariables(VVAccess.ReadWrite)]
+        [DataField("adminOnly")]
+        public bool AdminOnly { get; set; } = false;
+
+        [DataField("key", readOnly: true, required: true)]
+        private string _keyRaw = default!;
+
+        /// <summary>
+        ///     The client channel currently using the object, or null if there's none/not single user.
+        ///     NOTE: DO NOT DIRECTLY SET, USE ActivatableUISystem.SetCurrentSingleUser
+        /// </summary>
+        [ViewVariables]
+        public IPlayerSession? CurrentSingleUser;
+
+        void ISerializationHooks.AfterDeserialization()
+        {
+            var reflectionManager = IoCManager.Resolve<IReflectionManager>();
+            if (reflectionManager.TryParseEnumReference(_keyRaw, out var key))
+                Key = key;
+        }
+
+        public void OnPlayerStatusChanged(object? sender, SessionStatusEventArgs e)
+        {
+            // Has to be here for delegate to bind correctly.
+            // This *CANNOT* be elevated to system level because otherwise:
+            // pick one:
+            // + as method: delegate doesn't get the context it needs
+            //              (and will break if multiple show up)
+            // + as lambda: it can't be properly removed
+            // This probably ought to be BUI's responsibility anyway
+            // A disconnect should count as closing the UI, rest follows automatically
+            if (Deleted || e.Session != CurrentSingleUser || e.NewStatus != SessionStatus.Disconnected) return;
+            _activatableUISystem.SetCurrentSingleUser(OwnerUid, null, this);
+        }
+
+        protected override void Initialize()
+        {
+            base.Initialize();
+
+            // Note the implicit exception thrown if the key is messed up
+            // This is on purpose for fail-fast
+            var ui = UserInterface;
+            if (ui != null)
+                ui.OnClosed += UserInterfaceOnClosed;
+
+            _activatableUISystem = EntitySystem.Get<ActivatableUISystem>();
+        }
+
+        private void UserInterfaceOnClosed(IPlayerSession player)
+        {
+            if (player != CurrentSingleUser) return;
+            _activatableUISystem.SetCurrentSingleUser(OwnerUid, null, this);
+        }
```

Comment:
Instead of relying on `OnClosed`, I believe you could just subscribe to `BoundUIClosedEvent`. Also, if the BUI closed event is not fired when a user disconnects, it absolutely should be, so all of the `OnPlayerStatusChanged` stuff should be unnecessary? 

Then I don't think you would need any logic in `ActivatableUIComponent` (ignoring the enum stuff).

Reply:
Will switch to BoundUIClosedEvent.

Though in regards to:
"Also, if the BUI closed event is not fired when a user disconnects, it absolutely should be, so all of the OnPlayerStatusChanged stuff should be unnecessary?"
This doesn't actually solve the problem...

Reply:
(To clarify: It means I have to go and introduce an engine PR into this)

Reply:
> This doesn't actually solve the problem...

Then I don't understand what the problem is here. As far as I can tell, you simply use `OnPlayerStatusChanged` to detect when a client disconnects, and then treat that as if they had closed the UI? I checked, and BUI does in fact raise a `BoundUIClosedEvent` on disconnects.  So as far as I can tell all the `OnPlayerStatusChanged` stuff is unnecessary?

What is the engine PR you would need / what problem still needs to be solved?

Reply:
Wait, it *does?* Then why does the instrument code make a big deal of checking for disconnect...???

Reply:
Nevermind - I'll delete this whole block later today and test it.

Reply:
> Nevermind - I'll delete this whole block later today and test it.

The code is still there, so did it not work? It works fine when I test it without this code?

Reply:
oh, wait, is it? I honestly thought I removed it. I'll check.

Reply:
Removed, Rebased, Rechecked